File: src/main/java/edu/psu/ist/plato/kaiming/arm64/Operand.java
package edu.psu.ist.plato.kaiming.arm64;

public interface Operand {
    
    public enum Type {
        IMMEDIATE,
        REGISTER,
        MEMORY,
    }
    
    public boolean isRegister();
    
    public boolean isMemory();

    public boolean isImmeidate();
    
    public Type type();
    
    public default Immediate asImmediate() {
        return (Immediate)this;
    }
    
    public default Memory asMemory() {
        return (Memory)this;
    }
    
    public default Register asRegister() {
        return (Register)this;
    }
    
}

File: src/main/scala/edu/psu/ist/plato/kaiming/arm64/Operand.scala
package edu.psu.ist.plato.kaiming.arm64

trait Operand {
  
  def isRegister = this.isInstanceOf[Register]
  def isMemory = this.isInstanceOf[Memory]
  def isImmediate = this.isInstanceOf[Immediate]
  
  def asImmediate: Immediate = 
    throw new UnsupportedOperationException(this + "is not an immediate operand") 

  def asMemory: Memory = 
    throw new UnsupportedOperationException(this + "is not a memory operand") 
  
  def asRegister: Register = 
    throw new UnsupportedOperationException(this + "is not a register operand") 
  
}